### PR TITLE
Mouse picker branch

### LINKED ISSUE #190: Mouse picking
Select an object by clicking the mouse button.

### LINKED ISSUE #309: Mouse picking editor inside overlapping
As it is now, if you're inside an entity (like the example scene's 'cave') it will always override all other intersects unless it is checked last by putting it furthest down on the entity list in the editor.

Solution suggestion: Make it so that the program checks if you're inside an entity, and in that case, check that entity last.

### PR DESCRIPTION
**Pull request by SwedishSock**
Adds mouse picking in game, and fixes the earlier issue of picking the closest entity in the editorial mouse picking.

fixes #190 
fixes #309

### CODE REVIEW

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L507](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/bd0ce4d99dcd79a102e2255f0146625020e4f564/src/Engine/Manager/ScriptManager.cpp#L507): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L507](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/57cff3e2c5692598fcc7ca1a6bed935ff0243809/src/Engine/Manager/ScriptManager.cpp#L507): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Engine/Util/MousePicking.hpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include "Engine/Entity/Entity.hpp"
+
+/// Use mouse cursor to select objects in editor.
```
Comment (Chainsawkitten) [inline src/Engine/Util/MousePicking.hpp]: Also in game.

File: src/Editor/Editor.cpp [hunk 2]
```diff
@@ -245,20 +245,30 @@ void Editor::Show(float deltaTime) {
         if (Input()->Pressed(InputHandler::SELECT) && !ImGui::IsMouseHoveringAnyWindow()) {
             mousePicker.UpdateProjectionMatrix(cameraEntity->GetComponent < Component::Lens>()->GetProjection(glm::vec2(MainWindow::GetInstance()->GetSize().x, MainWindow::GetInstance()->GetSize().y)));
             mousePicker.Update();
-
-            for (int i = 0; i < Hymn().world.GetEntities().size(); ++i) {
+            float lastDistance = INFINITY;
+            int entityIndex = 0;
+            int entityAmount = Hymn().world.GetEntities().size();
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: MINOR The scope of the variable 'intersectDistance' can be reduced. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AvariableScope)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L485](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/5ec4c799fa5fa338993f5cb5f5d3a90e3aa200ad/src/Engine/Manager/ScriptManager.cpp#L485): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Editor/Editor.cpp [hunk 3]
```diff
@@ -245,20 +245,30 @@ void Editor::Show(float deltaTime) {
         if (Input()->Pressed(InputHandler::SELECT) && !ImGui::IsMouseHoveringAnyWindow()) {
             mousePicker.UpdateProjectionMatrix(cameraEntity->GetComponent < Component::Lens>()->GetProjection(glm::vec2(MainWindow::GetInstance()->GetSize().x, MainWindow::GetInstance()->GetSize().y)));
             mousePicker.Update();
-
-            for (int i = 0; i < Hymn().world.GetEntities().size(); ++i) {
-                float intersectDistance = 0.0f;
+            float lastDistance = INFINITY;
+            int entityIndex = 0;
+            int entityAmount = Hymn().world.GetEntities().size();
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: MINOR The scope of the variable 'intersectDistance' can be reduced. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AvariableScope)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L486](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/c761181b56d8c8a984643eb5a4e9154bb1d67338/src/Engine/Manager/ScriptManager.cpp#L486): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L486](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/3c660a6ef4fa5023c9aa5e6ed8373b1060d468f7/src/Engine/Manager/ScriptManager.cpp#L486): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L506](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/0f5c31afdc6380df4e5d9cdcd47392b516cd418a/src/Engine/Manager/ScriptManager.cpp#L506): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L506](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/f54f3d440557e668f541634eafe0311c2f4adb01/src/Engine/Manager/ScriptManager.cpp#L506): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**